[BugFix] Dictionary refresh maybe timeout if the dictionary object is huge (#39380)

srlch opened the pull request:
Why I'm doing:
If we refresh a dictionary cache using a huge dictionary table, refresh timeout may happen. The problem is that, we refresh 256MB chunk for a single chunk in a single refresh task. Each refresh task is executed concurrently. Since this chunk is relatively large, the respond time of a certain task's rpc request may be too long and cause a timeout.

What I'm doing:
1. Make refresh cache rpc timeout configable.
2. Make chunk smaller when it used for refresh task. Helps reduce rpc response time.
3. Make the refresh task thread pool size configable. By controlling concurrency, speed and task pressure can be balanced.
4. Unlimit the queue size of task thread pool. Because tasks are divided into smaller pieces, the number of tasks may increase.

Fixes #39380

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

trueeyu commented on be/src/common/config.h:
```diff
@@ -1173,4 +1173,7 @@ CONF_mInt64(lake_local_pk_index_unused_threshold_seconds, "86400"); // 1 day
 
 CONF_mBool(lake_enable_vertical_compaction_fill_data_cache, "false");
 
+CONF_Int32(dictionary_cache_refresh_timeout, "60000"); // 1 min
```
_ms ?

srlch replied:
yes

trueeyu replied:
```
eyu@eyu:~/code/starrocks$ grep "_timeout" be/src/common/config.h
CONF_Int32(thrift_connect_timeout_seconds, "3");
CONF_Int32(broker_write_timeout_seconds, "30");
CONF_Int32(tablet_writer_open_rpc_timeout_sec, "60");
CONF_Int32(make_snapshot_rpc_timeout_ms, "20000");
// Deprecated, use query_timeout instread
// CONF_Int32(tablet_writer_rpc_timeout_sec, "600");
CONF_mInt32(thrift_rpc_timeout_ms, "5000");
CONF_mInt32(txn_commit_rpc_timeout_ms, "60000");
CONF_Int32(routine_load_kafka_timeout_second, "10");
CONF_Int32(routine_load_pulsar_timeout_second, "10");
CONF_Int32(es_http_timeout_ms, "5000");
CONF_Int64(object_storage_connect_timeout_ms, "-1");
CONF_Int64(object_storage_request_timeout_ms, "-1");
CONF_Int64(send_rpc_runtime_filter_timeout_ms, "1000");
CONF_Int64(rpc_connect_timeout_ms, "30000");
CONF_mInt32(starlet_fslib_s3client_connect_timeout_ms, "1000");
CONF_Int32(jdbc_connection_idle_timeout_ms, "600000");
CONF_mInt64(load_tablet_timeout_seconds, "60");

```

srlch replied:
done

srlch committed:
[BugFix] Dictionary refresh maybe timeout if the dictionary object is huge (#39380)

Why I'm doing:
If we refresh a dictionary cache using a huge dictionary table, refresh timeout may
happen. The problem is that, we refresh 256MB chunk for a single chunk in a single
refresh task. Each refresh task is executed concurrently. Since this chunk is
relatively large, the respond time of a certain task's rpc request may be too long
and cause a timeout.

What I'm doing:
1. Make refresh cache rpc timeout configable.
2. Make chunk smaller when it used for refresh task. Helps reduce rpc response time.
3. Make the refresh task thread pool size configable. By controlling concurrency, speed
and task pressure can be balanced.
4. Unlimit the queue size of task thread pool. Because tasks are divided into smaller pieces,
the number of tasks may increase.

Signed-off-by: srlch <linzichao@starrocks.com>

trueeyu commented on be/src/common/config.h:
```diff
@@ -1173,4 +1173,7 @@ CONF_mInt64(lake_local_pk_index_unused_threshold_seconds, "86400"); // 1 day
 
 CONF_mBool(lake_enable_vertical_compaction_fill_data_cache, "false");
 
+CONF_Int32(dictionary_cache_refresh_timeout_ms, "60000"); // 1 min
```
Can dynamic modification be supported?

trueeyu commented on be/src/common/config.h:
```diff
@@ -1173,4 +1173,7 @@ CONF_mInt64(lake_local_pk_index_unused_threshold_seconds, "86400"); // 1 day
 CONF_mBool(lake_enable_vertical_compaction_fill_data_cache, "false");
 
+CONF_Int32(dictionary_cache_refresh_timeout_ms, "60000"); // 1 min
+CONF_Int32(dictionary_cache_refresh_threadpool_size, "8");
```
Can dynamic modification be supported?

srlch replied:
done

srlch committed:
Make configable

Signed-off-by: srlch <linzichao@starrocks.com>
